### PR TITLE
Introduce new API for setting file permissions

### LINKED ISSUE #21171: FileCopyDetails.setMode should have a better API for Kotlin DSL
`FileCopyDetails.setMode` currently forces us to provide the file mode as an int.

But the value the docs tell us to pass in as an example is in octal, and Kotlin doesn't support octal literals. So we have to jump through hoops to pass in an appropriate value.

### Expected Behavior

A method like `mode(action: FileModeDetails.() -> Unit)` could pass a details object into the action, which then has a number of methods available to set the file mode. After the action returns, it can convert this back to an int, if that's how it prefers to work with file modes internally.

So we could then write something like:

```kotlin
        if (path.startsWith("bin/")) {
            mode {
                executable = true
            }
        }
```

Or `setExecutableForAll(true)`, or whatever.

### Current Behavior

Currently we're forced to either provide the file mode in decimal or hex, or to use a string and convert it ourselves at runtime.

```kotlin
        if (path.startsWith("bin/")) {
            mode = "755".toInt(radix = 8)
        }
```

But even this doesn't clearly describe the intent. The reader doesn't necessarily know that this is adding the ability to execute. Maybe they don't know that it's a UNIX file mode without reading the docs. Or maybe I'm setting the mode to _reduce_ permissions on the file.

### Context

Files packed inside a jar tend to have lost their modes. This means if you have a jar which contains executable files, you have to do something like this:

```kotlin
val copyStuff by tasks.registering(Copy::class) {
    from(zipFilePath)
    eachFile {
        if (path.startsWith("bin/")) {
            mode = "755".toInt(radix = 8)
        }
    }
}
```

Kotlin themselves seem to think that nobody should be using octal anymore, so it's highly unlikely they will ever add octal literals.

Reference: [Kotlin feature request for octal literals (declined)](https://youtrack.jetbrains.com/issue/KT-44409).

### PR DESCRIPTION
**Pull request by jbartok**
Fixes #21171

### CODE REVIEW

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java
Comment (jbartok) [inline subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java:38]: `@Input`

**Review by tresat — APPROVED** [review]

File: subprojects/core-api/src/main/java/org/gradle/api/file/ReadOnlyFileAccessPermissions.java
Comment (eskatos) [reply subprojects/core-api/src/main/java/org/gradle/api/file/ReadOnlyFileAccessPermissions.java:28]: +1 to `FileAccessPermissions` and `MutableFileAccessPermissions`

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileTreeElement.java
Comment (eskatos) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileTreeElement.java:112]: Maybe `getImmutablePermissions()`, aligned with using "mutable" in the other type name (see other comment)

File: subprojects/core-api/src/main/java/org/gradle/api/internal/file/ReadOnlyFileAccessPermissionInternal.java
Comment (eskatos) [reply subprojects/core-api/src/main/java/org/gradle/api/internal/file/ReadOnlyFileAccessPermissionInternal.java:29]: +1, especially with the deprecation of the `getMode()` functions

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java:42]: The read-only version of `FileAccessPermission` has the same flags as the non-read-only version, ie. read, write & execute. It's read-only due to the fact that you can only read those flags and not set them. But, if this confuses you, then it might confuse others too, so will think about it. Anyways, the fate of this read-only thing is not decided yet.

Comment (tresat) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java:44]: It's not a big deal, but I do think if this becomes `MutableFileAccessPermission` -extends-> `FileAccessPermission` it's just clearer all around.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CopyPermissionsIntegrationTest.groovy
Comment (jbartok) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CopyPermissionsIntegrationTest.groovy:479]: :) unless you need multiple such columns...

File: subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java:95]: :thought_balloon: I think there's an unnecessary `Provider` creation here
```suggestion
            getInputs().property(specPropertyName + ".dirPermissions", spec.getDirPermissions().map(ImmutableFileAccessPermissions::toUnixNumeric))
```
This change requires changing imports

File: subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java:97]: :thought_balloon: I think there's an unnecessary `Provider` creation here
```suggestion
            getInputs().property(specPropertyName + ".filePermissions", spec.getFilePermissions().map(ImmutableFileAccessPermissions::toUnixNumeric))
```
This change requires changing imports

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/ApplicationPlugin.java
Comment (eskatos) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/ApplicationPlugin.java:227]: Or the symbolic notation which this more readable.

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java
Comment (eskatos) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileAccessPermission.java:40]: If one wants to use the permissions as an input they can use `toUnixNumeric() as a declared input.
This type `FileAccessPermission` and `FileAccessPermissions` don't need input annotations.

File: subprojects/core-api/src/main/java/org/gradle/api/file/ImmutableFileAccessPermissions.java
Comment (eskatos) [inline subprojects/core-api/src/main/java/org/gradle/api/file/ImmutableFileAccessPermissions.java:41]: :x: This should return a `Provider<Integer>` instead, so it doesn't realize lazy values too early and can be used as a task input while retaining possible dependencies.

**Review by eskatos — CHANGES_REQUESTED** [review]
Good work. Test coverage, error reporting and the API are nice.

I requested a couple changes and made remarks here and there.

----

Documentation is missing. This can be addressed in a subsequent PR.

It would be nice if the javadoc of `FileAccessPermissions` would include a snippet showing how the API is used.

This needs to be in the user manual with code snippets. This could be added to `Working With Files` in the `File copying in depth` after [Filtering file content](https://docs.gradle.org/current/userguide/working_with_files.html#sec:filtering_files) and before [Using the CopySpec class](https://docs.gradle.org/current/userguide/working_with_files.html#sec:using_the_copyspec_class).


The Groovy DSL reference needs to be updated, e.g. https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:fileMode

This is release notes worthy!

Release notes and upgrading guide should link to the user manual.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileSystemOperations.java
Comment (jbartok) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileSystemOperations.java:34]: I'm not lazy to put one there, but why would we? If one is needed, wouldn't a test fail somewhere? Please explain what I'm missing.

File: subprojects/core-api/src/main/java/org/gradle/api/file/ImmutableFileAccessPermissions.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/ImmutableFileAccessPermissions.java:41]: Agree on principle, but I'm getting terribly ugly internal usages, like:

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java:166]: We should strive at not using `Providers.of()` when possible in order to not break provider chains.

It looks like this could be replaced with:
```java
permissions.flatMap(ImmutableFileAccessPermissions::toUnixNumeric).map(DefaultImmutableFileAccessPermissions::new)
```

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java:174]: Same remark about unnecessary usage of `Providers.of()`

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractImmutableFileAccessPermission.java
Comment (eskatos) [inline subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractImmutableFileAccessPermission.java:35]: This is breaking the chain of providers. I think this should return a `Provider<Integer>`.

```suggestion
    protected Provider<Integer> toUnixNumeric() {
        return getRead().map(r -> r ? 4 : 0)
            .zip(getWrite().map(w -> w ? 2 : 0), Integer::sum)
            .zip(getExecute().map(x -> x ? 1 : 0), Integer::sum);
    }
```

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractImmutableFileAccessPermissions.java
Comment (eskatos) [inline subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractImmutableFileAccessPermissions.java:34]: Following up on the provider chain breakage in `AbstractImmutableFileAccessPermission` above, here's how I think this should be implemented:

```suggestion
    public Provider<Integer> toUnixNumeric() {
        AbstractImmutableFileAccessPermission user = (AbstractImmutableFileAccessPermission) getUser();
        AbstractImmutableFileAccessPermission group = (AbstractImmutableFileAccessPermission) getGroup();
        AbstractImmutableFileAccessPermission other = (AbstractImmutableFileAccessPermission) getOther();
        return user.toUnixNumeric().map(u -> 64 * u)
            .zip(group.toUnixNumeric().map(g -> 8 * g), Integer::sum)
            .zip(other.toUnixNumeric(), Integer::sum);
    }
```

**Comment by jbartok** [comment]
@bot-gradle test this

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/ShadowPluginSmokeTest.groovy
Comment (eskatos) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/ShadowPluginSmokeTest.groovy:53]: We are testing the latest version of shadow.
This is not a "legacy" deprecation warning.
We can't ship with such nagging that users can't do anything about.

What we should do is the following:
* remove the nagging for now
* open issues with third parties plugin to move to the new API once available
* open issues on our side to add the nagging once there are releases of fixed 3rd parties available with links to these issues, and schedule them for the next milestone for a revisit

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/SpringBootPluginSmokeTest.groovy
Comment (eskatos) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/SpringBootPluginSmokeTest.groovy:89]: Here we don't test the latest Spring Boot version.

* See https://github.com/gradle/gradle/issues/23937

But the latest will still be using the old API.

We should follow the same process as for Shadow.

**Review by eskatos — CHANGES_REQUESTED** [review]

**Review by eskatos — APPROVED** [review]

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Comment by jbartok** [comment]
@bot-gradle help

**Comment by jbartok** [comment]
@bot-gradle test APT

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test APT

File: subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java]: https://github.com/gradle/gradle/pull/24919

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

**Review by eskatos — APPROVED** [review]
:+1:

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test APT

**Review by eskatos — APPROVED** [review]

**Comment by jbartok** [comment]
@bot-gradle test and merge

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt
Comment (lptr) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:503]: ❓ Why not add the annotation instead?

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt
Comment (lptr) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:6524]: :x: What is the difference here?

File: subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java:92]: ❓ This method is not deprecated, yet we say "Use X instead". Feels incorrect.

**Review by lptr — COMMENTED** [review]
I left a few random comments, but the main issue is with the use of lazy vs eager properties and return values. I think it would be good to discuss the best way to approach the problem in a spec first.

To me it looks like we have some aspects to this:

- `FileCopyDetails` should only expose eager values, as it is only present during execution where lazy configuration doesn't make sense.

- The permissions should be set lazily during configuration on `CopySpec`s. However, this can be achieved in different ways, and we don't need to provide laziness on every level. Either `FileAccessPermissions` should be lazy, and it should have eager `FileAccessPermission` properties (which would also have eager sub-properties for read-write-execute); or `FileAccessPermissions` should be a read-only eager singleton, and its user-group-other properties should be lazy, and then again, `FileAccessPermission` should only have eager properties. Or both `FileAccessPermissions` and its `FileAccessPermission` properties should be eager read-only singletons, and the `read-write-execute` sub-properties should be lazy. I have no strong feelings here about which way the design should go, perhaps the configuration team could be consulted?

- I also feel that the names `FileAccessPermission` and `FileAccessPermissions` are way too close to be confusing. Something further apart would be much better.

File: subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java:92]: Good catch. It's a leftover from having deprecations added in this PR but then pulling them out in a separate PR.

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt
Comment (jbartok) [reply subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt]: No idea what this was, went away after merging in `master`.

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test APT

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt
Comment (lptr) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:90]: ❓ Should we have a public API to set the permissions directly using an integer? It seems that we are now converting to-and-from a `String` to go in between.

File: subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/file/CopyProcessingSpec.java:143]: 💭  We really shouldn't add these methods manually. They should be generated instead. Not something to fix in this PR, but we should address this in a more general way.

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileCopyDetails.java
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/file/FileCopyDetails.java:66]: ❓ I think we shouldn't say "use X instead" for APIs that are not deprecated. Or we should deprecate them. We could use `@see` or something similar to point people to the new APIs perhaps. WDYT?

**Review by lptr — COMMENTED** [review]
I think the API needs a bit more improvement.

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemOperations.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemOperations.java:107]: There is actually a good reason for that, but there should be javadoc explaining it, so will add that.

File: subprojects/core-api/src/main/java/org/gradle/api/internal/file/UserClassFilePermissionsInternal.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/internal/file/UserClassFilePermissionsInternal.java:31]: Removing the whole thing. I don't like it either and the original reason for picking this form has been lost in refactorings.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt
Comment (jbartok) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:62]: Definitely not.

Comment (lptr) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:62]: Why not?

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt
Comment (jbartok) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:90]: yes

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileCopyDetails.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileCopyDetails.java:66]: Yeah, very last of the leftovers of moving deprecations to separate PR. Removing it.

Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileCopyDetails.java:66]: But, beyond that, I do like the idea of starting to point people to the new API, even while incubating.... "Consider using XXX instead?" Will give it a try...

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemOperations.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemOperations.java:107]: Ahh, that Javadoc is already there, explaining how these methods are not changing existing permissions, but overwriting them entirely, so it no longer matters that files and directories start out with different defaults.

File: subprojects/core-api/src/main/java/org/gradle/api/file/FilePermissions.java
Comment (jbartok) [reply subprojects/core-api/src/main/java/org/gradle/api/file/FilePermissions.java]: Ok, will rename like this.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt
Comment (jbartok) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:62]: Because what you suggest is about permissions for ONE SPECIFIC class of user and what we need here is all permissions for all classes of users.

Comment (lptr) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:62]: Maybe we are misunderstanding each other. What I mean is it would be better to use a method handle instead of a closure. Is there any reason not to do that?

**Comment by jbartok** [comment]
@lptr I think your comments are all addressed (and the outcome is definitely better than I had before). Please take another look when you can.

I would like to merge this in and then, if it needs further changes, do those in a new PR.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt
Comment (jbartok) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/DefaultCopySpecCodec.kt:62]: Ahh, that definitely is ok. The class name is wrong though in the suggestion, that's what I was arguing against. Lost in translation, will fix.

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**
